Add capability to stop directly on LINEAR the Terminator

GiudGiud opened the pull request:
@dschwen thoughts?

it s a tad violent to throw immediately but it does save time to not wait for a bad convergence history to run its course (the use case here: https://github.com/idaholab/moose/discussions/24457#discussioncomment-5989722)

As a reminder, the SOFT fail mode finishes the time step before refusing it as the fixedpointsolve has been set to fail the step by the Terminator. 

EDIT:
not throwing anymore

GiudGiud commented:
@dschwen 

should I use what you have here instead?
https://github.com/idaholab/moose/pull/23011/files

dschwen commented:
> Not that this is still a tad slower than the user asked for, since they wanted to stop on LINEAR

Really? But no further residual evaluations are performed once the failure flag is set, so the linear iterations take zero time.

GiudGiud commented:
ah awesome. 

Ok i think I should move to that. This will be much lower footprint than throwing an exception.

Unless you think there's some value in throwing an exception

GiudGiud commented:
Doesnt work, keeps moving forward

GiudGiud committed:
Add test for terminator on LINEAR
refs #24479

GiudGiud committed:
Allow Terminator soft fail on LINEAR
refs #24479

GiudGiud commented:
Nevermind, got to look more

GiudGiud commented:
a [[nodiscard]] would have saved me here haha

GiudGiud committed:
Rename poorly named failing API, using an action verb for a getter
follows up on #23011

GiudGiud committed:
Check for the postprocessors in the criterion expression for being up to date on evaluation
refs #24479

GiudGiud committed:
Add test for PP execution scheudle checking
refs #24479

GiudGiud commented:
SHould be ready

lindsayad commented on framework/include/problems/FEProblemBase.h:
```diff
@@ -2047,10 +2047,21 @@ class FEProblemBase : public SubProblem, public Restartable
    * Skip further residual evaluations and fail the next nonlinear convergence check
    */
-  bool failNextNonlinearConvergenceCheck() const { return _fail_next_nonlinear_convergence_check; }
+  void setFailNextNonlinearConvergenceCheck(bool fail)
+  {
+    _fail_next_nonlinear_convergence_check = fail;
```
Will this ever be called with a `false` argument? If not, I'd remove the argument

GiudGiud replied:
that s what I expected the `failNextNonlinearConvergenceCheck` to do, which lost me half an hour or so

the current configuration is the most explicit

GiudGiud replied:
Unfailing the solve. Heroic recovery

my concern now is what is people think that fail(false) means it will always not fail

GiudGiud replied:
changed

lindsayad commented on framework/src/userobjects/Terminator.C:
```diff
@@ -85,6 +85,24 @@ Terminator::Terminator(const InputParameters & parameters)
+{
+  // Check execution schedule of the postprocessors
+  if (_fail_mode == FailMode::SOFT)
+    for (unsigned int i = 0; i < _pp_num; i++)
```
```suggestion
    for (const auto i : make_range(_pp_num))
```

lindsayad commented on framework/src/userobjects/Terminator.C:
```diff
@@ -85,6 +85,24 @@ Terminator::Terminator(const InputParameters & parameters)
 }
 
+void
+Terminator::initialize()
```
Could this be run on `initialSetup` instead of `initialize`?

GiudGiud replied:
oh yeah

lindsayad commented on test/tests/userobjects/Terminator/tests:
```diff
@@ -6,14 +6,27 @@
+    cli_args = "Outputs/file_base=terminator_soft_solve_out
+                Postprocessors/dt/execute_on=LINEAR
+                UserObjects/arnold/execute_on=LINEAR"
+    requirement = "The system shall abort the current solve in a solve based on post-processing "
```
Redundant 'solve'

GiudGiud replied:
I ll fix the whole file

lindsayad commented on test/tests/userobjects/Terminator/tests:
```diff
@@ -25,7 +38,8 @@
   [terminator_message_2]
     type = 'RunException'
-    expect_err = "Setting the error level to ERROR always causes a hard failure, which is incompatible with `fail_mode=SOFT`."
+    expect_err = "Setting the error level to ERROR always causes a hard failure, which is "
+                 "incompatible with `fail_mode=SOFT`."
```
```suggestion
    expect_err = "Setting the error level to error always causes a hard failure, which is "
                 "incompatible with a soft failure mode."
```
Here I am introducing my own redundancy but I don't really know how not to

GiudGiud replied:
ERROR and fail_mode=SOFT are in the parameters.
I think it s clearer how it used to be.

undid the change

lindsayad replied:
Requirements are supposed to be independent of implementation. The requirement is conceptually supposed to occur before the implementation. This concretely shows implementation

GiudGiud replied:
This isnt a requirement though. It s an expected error message

lindsayad replied:
Oh my goodness. I'm a total dummy. I'm blind as a bat!

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1550485/) on 6e371fc wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/24463/site)

This comment will be updated on new commits.

GiudGiud committed:
Apply suggestions from code review
- better tests requirement
- dont allow for recovering the failed solve boolean
- better for loop
- remmber that initialize is not the same as initialSetup

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

GiudGiud commented:
Review should be fully addressed.

GiudGiud commented:
I ll make all my newsletter contribution in one single PR btw. Not trying to add this here

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1551412/) on 5cc5fa7 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>4f29f1</th>
      <th colspan=4>#24463 5cc5fa</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.37%</td>
      <td>85.38%</td>
      <td>+0.00%</td>
      <td>94.74%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>88342</td>
      <td>88359</td>
      <td>+17</td>
      <td>18</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15137</td>
      <td>15136</td>
      <td>-1</td>
      <td>1</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24463/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24463/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/24463/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/24463/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/24463/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/24463/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/24463/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/24463/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/24463/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/24463/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/24463/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/24463/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/24463/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/24463/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/24463/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/24463/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/24463/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/24463/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/24463/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/24463/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/24463/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/24463/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/24463/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/24463/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/24463/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/24463/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/24463/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/24463/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/24463/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/24463/coverage/xfem)

</details>



This comment will be updated on new commits.

GiudGiud commented:
Green alas

lindsayad reviewed (approved):
I will actually approve this since you are only adding a new line to the existing requirement. But I still think the requirement is not well written

GiudGiud commented:
Both recipes passed earlier with no changes in between. 
The doc failure and openmpi are unrelated

GiudGiud commented:
Thanks for the review!
